### PR TITLE
[NEW] Better Push and Email Notification logic

### PR DESCRIPTION
**Pull request by rodrigok**
## Changelog Description

We are still using the same logic to define which notifications every new message will generate, it takes some servers' settings, users's preferences and subscriptions' settings in consideration to determine who will receive each notification type (desktop, audio, email and mobile push), but now it doesn't check the user's status (online, away, offline) for email and mobile push notifications but send those notifications to a new queue with the following rules:

- When the user is online the notification is scheduled to be sent in 120 seconds
- When the user is away the notification is scheduled to be sent in 120 seconds minus the amount of time he is away
- When the user is offline the notification is scheduled to be sent right away
- When the user reads a channel all the notifications for that user are removed (clear queue)
- When a notification is processed to be sent to a user and there are other scheduled notifications:
  - All the scheduled notifications for that user are rescheduled to now
  - The current notification goes back to the queue to be processed ordered by creation date

## Technical info
`maxScheduleDelaySeconds = 120`

- [x] Create a description for this PR
- [x] Merge [base PR](https://github.com/RocketChat/Rocket.Chat/pull/17338) and change the target merge branch here to develop
- [x] Make `cyclePause`, `maxBatchSize`,  and `maxScheduleDelaySeconds` configurable
- [x] if connection online always schedule for `maxScheduleDelaySeconds` time
- [x] if connection away schedule for min(`maxScheduleDelaySeconds` - (now - lastUpdatedAt), 0)
- [x] if connection offline schedule for 0
- [x] on a notification remove all schedules of other user's notification to flush the queue
- [x] If user read a room, clear all the scheduled notifications for the user
- [x] Create indexes
- [x] Remove the push queue
- [x] TTL to remove old records (2h maybe?)
- [x] Migration to remove `Notifications_Always_Notify_Mobile` and `Push_Debug`
- [x] Migrate and drop the `_raix_push_notifications` collection
  - [x] Migrate `query.userId` to `userId`

# Future
- We may improve to create a tick to set user's activity later (from the client) and use this field to calculate the schedule and to clear the queue
- Set push tokens inside the session token

### CODE REVIEW

File: app/notification-queue/server/NotificationQueue.ts [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+import { NotificationQueue, Users } from '../../models/server/raw';
+import { sendEmailFromData } from '../../lib/server/functions/notifications/email';
+import { PushNotification } from '../../push-notifications/server';
+
+const UsersCollection: Collection = Users.col;
+
+interface INotificationItemPush {
+	type: 'push';
+	data: {
+		roomId: string;
```
Comment (sampaiodiego) [inline app/notification-queue/server/NotificationQueue.ts:13]: why is this information duplicated from `INotification`?

if really needed, why not use `rid` and `uid` as used there?

File: server/startup/migrations/v185.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+			}],
+		};
+	} catch (e) {
+		//
+	}
+}
+
+Migrations.add({
+	version: 185,
+	async up() {
```
Comment (sampaiodiego) [inline server/startup/migrations/v185.js:41]: if you want to hold until this migration is complete, you'll need to use `Promise.await()` , using `async` here will return a promise right away and the migration will process continue.

you can look how I did on migration 179:

https://github.com/RocketChat/Rocket.Chat/blob/develop/server/startup/migrations/v179.js

File: app/notification-queue/server/NotificationQueue.ts (hunk 1 above)
Comment (rodrigok) [reply app/notification-queue/server/NotificationQueue.ts:13]: I didn't change the content of the push and email, just saving them and reading.

File: app/models/server/raw/BaseRaw.js [hunk 3]
```diff
@@ -19,8 +21,11 @@ export class BaseRaw {
 		return this.col.find(...args);
 	}
 
-	insert(...args) {
-		return this.col.insert(...args);
+	insert(docs, options) {
```
[the remark this answers is not in the document]
Comment (rodrigok) [reply app/models/server/raw/BaseRaw.js:24]: I'm not using this method anymore, and I didn't change because I'm not sure where else it's used

**Comment by ankar84** [comment]
I think that PR could be a great fix for
https://github.com/RocketChat/Rocket.Chat/issues/16867
https://github.com/RocketChat/Rocket.Chat/issues/16765
@rodrigok @sampaiodiego thank you guys!
That push notification algorithm is much straightforward and I believe convenient to rocket chat users.
Thanks a lot again!

**Comment by sampaiodiego** [comment]
thanks @ankar84 .. I believe both issues can be closed now, what do you think? even the preference `Always notify mobile` was removed.

**Comment by ankar84** [comment]
> I believe both issues can be closed now, what do you think?

Sure! They both closed now.
Thanks again!

**Comment by ceefour** [comment]
@rodrigok @sampaiodiego I'd like to confirm behavior or following scenario:

1. App is online/foreground
2. Server has message to be posted, meaning push is scheduled in 120 seconds
3. Before message is received by app, and before 120 seconds timed out, app is now offline *and server knows the status is now offline* (perhaps away also applies here)

Since the status is now (known to be) offline, the best user experience means that push needs to be sent immediately, without waiting for the scheduled 120 seconds.

Is that already the current behavior?

My observation seems to imply that, even now the user is known to be offline, all scheduled messages will still be waiting until 120 seconds.

**Pull request merged**